zerver/lib/user_groups.py

```diff
@@ -45,17 +67,55 @@ def access_user_group_by_id(
     return user_group
 
 
-def access_user_groups_as_potential_subgroups(
-    user_group_ids: Sequence[int], acting_user: UserProfile
-) -> List[UserGroup]:
-    user_groups = UserGroup.objects.filter(id__in=user_group_ids, realm=acting_user.realm)
+@contextmanager
+def lock_subgroups_with_respect_to_supergroup(
+    potential_subgroup_ids: Collection[int], potential_supergroup_id: int, acting_user: UserProfile
+) -> Iterator[LockedUserGroupContext]:
+    """This locks the user groups with the given potential_subgroup_ids, as well as their indirect subgroups,
+    followed by the potential supergroup. It ensures that we lock the user groups in a consistent order
+    topologically to avoid unnecessary deadlocks on non-conflicting queries.
 
-    valid_group_ids = [group.id for group in user_groups]
-    invalid_group_ids = [group_id for group_id in user_group_ids if group_id not in valid_group_ids]
-    if invalid_group_ids:
-        raise JsonableError(_("Invalid user group ID: {}").format(invalid_group_ids[0]))
+    Regardless of whether the user groups returned are used, always call this helper before making
+    changes to subgroup memberships. This avoids introducing cycles among user groups when there is a
+    race condition in which one of these subgroups become an ancestor of the parent user group in
+    another transaction.
 
-    return list(user_groups)
+    Note that it only does the access_user_group_by_id permission check on the potential supergroup, not
+    the potential subgroups and their recursive subgroups. However, the permission model for accessing subgroups
+    is reserved for changes.
+    """
+    with transaction.atomic():
```

Comment:
`durable=True` cannot be used here because we use `compose_view` to combine multiple such blocks together. However, I'm not 100% sure if `savepoint=False` should be used.

Reply:
I think `savepoint=False` is correct. We never want to use savepoints, as they have a bunch of problematic quirks.